src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,633 @@
+    return 1;
+}
+
+/* Normalizes the x-coordinate of the given group element. */
+static int secp256k1_xonly_ge_serialize(unsigned char *output32, secp256k1_ge *ge) {
+    if (secp256k1_ge_is_infinity(ge)) {
+        return 0;
+    }
+    secp256k1_fe_normalize_var(&ge->x);
+    secp256k1_fe_get_b32(output32, &ge->x);
+    return 1;
+}
+
+static void secp256k1_nonce_function_musig(secp256k1_scalar *k, const unsigned char *session_id, const unsigned char *key32, const unsigned char *msg32, const unsigned char *agg_pk32, const unsigned char *extra_input32) {
+    secp256k1_sha256 sha;
+    unsigned char seed[32];
+    unsigned char i;
+    enum { n_extra_in = 4 };
+    const unsigned char *extra_in[n_extra_in];
+
+    /* TODO: this doesn't have the same sidechannel resistance as the BIP340
+     * nonce function because the seckey feeds directly into SHA. */
+
+    /* Subtract one from `sizeof` to avoid hashing the implicit null byte */
+    secp256k1_sha256_initialize_tagged(&sha, (unsigned char*)"MuSig/nonce", sizeof("MuSig/nonce") - 1);
+    secp256k1_sha256_write(&sha, session_id, 32);
+    extra_in[0] = key32;
+    extra_in[1] = agg_pk32;
+    extra_in[2] = msg32;
+    extra_in[3] = extra_input32;
+    for (i = 0; i < n_extra_in; i++) {
+        unsigned char marker;
+        if (extra_in[i] != NULL) {
+            marker = 1;
+            secp256k1_sha256_write(&sha, &marker, 1);
+            secp256k1_sha256_write(&sha, extra_in[i], 32);
+        } else {
+            marker = 0;
+            secp256k1_sha256_write(&sha, &marker, 1);
+        }
+    }
+    secp256k1_sha256_finalize(&sha, seed);
+
+    for (i = 0; i < 2; i++) {
+        unsigned char buf[32];
+        secp256k1_sha256_initialize(&sha);
+        secp256k1_sha256_write(&sha, seed, 32);
+        secp256k1_sha256_write(&sha, &i, sizeof(i));
+        secp256k1_sha256_finalize(&sha, buf);
+        secp256k1_scalar_set_b32(&k[i], buf, NULL);
+    }
+}
+
+int secp256k1_musig_nonce_gen(const secp256k1_context* ctx, secp256k1_musig_secnonce *secnonce, secp256k1_musig_pubnonce *pubnonce, const unsigned char *session_id32, const unsigned char *seckey, const unsigned char *msg32, const secp256k1_musig_keyagg_cache *keyagg_cache, const unsigned char *extra_input32) {
+    secp256k1_keyagg_cache_internal cache_i;
+    secp256k1_scalar k[2];
+    secp256k1_ge nonce_pt[2];
+    int i;
+    unsigned char pk_ser[32];
+    unsigned char *pk_ser_ptr = NULL;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secnonce != NULL);
+    memset(secnonce, 0, sizeof(*secnonce));
+    ARG_CHECK(pubnonce != NULL);
+    memset(pubnonce, 0, sizeof(*pubnonce));
+    ARG_CHECK(session_id32 != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+
+    /* Check that the seckey is valid to be able to sign for it later. */
+    if (seckey != NULL) {
+        secp256k1_scalar sk;
+        int ret;
+        ret = secp256k1_scalar_set_b32_seckey(&sk, seckey);
```

Comment:
Yes, that would help. I accidentally removed this line in a fixup yesterday. I pushed a commit that gets rid of the declassify entirely as @real-or-random suggested because that's consistent with ecdsa_sign for example.

"be able to sign for it later" means that due to the check you won't run into a situation where you generated a nonce with a secret key that won't work in partial_sign later because it's out of range.